File: app/src/main/java/wangdaye/com/geometricweather/basic/model/weather/MoonPhase.java
package wangdaye.com.geometricweather.basic.model.weather;

import android.content.Context;
import android.text.TextUtils;

import androidx.annotation.Nullable;

import wangdaye.com.geometricweather.R;

/**
 * Moon phase.
 * */
public class MoonPhase {

    @Nullable private Integer angle;
    @Nullable private String description;

    public MoonPhase(@Nullable Integer angle, @Nullable String description) {
        this.angle = angle;
        this.description = description;
    }

    @Nullable
    public Integer getAngle() {
        return angle;
    }

    @Nullable
    public String getDescription() {
        return description;
    }

    public boolean isValid() {
        return angle != null && description != null;
    }

    @Nullable
    public String getMoonPhase(Context context) {
        if (TextUtils.isEmpty(description)) {
            return context.getString(R.string.phase_new);
        }

        switch (description.toLowerCase()) {
            case "waxingcrescent":
            case "waxing crescent":
                return context.getString(R.string.phase_waxing_crescent);

            case "first":
            case "firstquarter":
            case "first quarter":
                return context.getString(R.string.phase_first);

            case "waxinggibbous":
            case "waxing gibbous":
                return context.getString(R.string.phase_waxing_gibbous);

            case "full":
            case "fullmoon":
            case "full moon":
                return context.getString(R.string.phase_full);

            case "waninggibbous":
            case "waning gibbous":
                return context.getString(R.string.phase_waning_gibbous);

            case "third":
            case "thirdquarter":
            case "third quarter":
            case "last":
            case "lastquarter":
            case "last quarter":
                return context.getString(R.string.phase_third);

            case "waningcrescent":
            case "waning crescent":
                return context.getString(R.string.phase_waning_crescent);

            default:
                return context.getString(R.string.phase_new);
        }
    }
}

File: app/src/main/java/org/breezyweather/common/basic/models/weather/MoonPhase.kt
package org.breezyweather.common.basic.models.weather

import android.content.Context
import org.breezyweather.R
import org.shredzone.commons.suncalc.MoonPhase
import java.io.Serializable
import java.util.Locale

/**
 * Moon phase.
 */
class MoonPhase(
    /**
     * Angle between 0 to 360 (no negative! Add 180 if you have negative numbers)
     */
    val angle: Int? = null
) : Serializable {

    val isValid: Boolean
        get() = angle != null

    fun getDescription(context: Context): String? {
        if (angle == null) return null

        return when (MoonPhase.Phase.toPhase(angle.toDouble())) {
            MoonPhase.Phase.NEW_MOON -> context.getString(R.string.ephemeris_moon_phase_new_moon)
            MoonPhase.Phase.WAXING_CRESCENT -> context.getString(R.string.ephemeris_moon_phase_waxing_crescent)
            MoonPhase.Phase.FIRST_QUARTER -> context.getString(R.string.ephemeris_moon_phase_first_quarter)
            MoonPhase.Phase.WAXING_GIBBOUS -> context.getString(R.string.ephemeris_moon_phase_waxing_gibbous)
            MoonPhase.Phase.FULL_MOON -> context.getString(R.string.ephemeris_moon_phase_full_moon)
            MoonPhase.Phase.WANING_GIBBOUS -> context.getString(R.string.ephemeris_moon_phase_waning_gibbous)
            MoonPhase.Phase.LAST_QUARTER -> context.getString(R.string.ephemeris_moon_phase_last_quarter)
            MoonPhase.Phase.WANING_CRESCENT -> context.getString(R.string.ephemeris_moon_phase_waning_crescent)
            else -> null
        }
    }

    companion object {
        fun getAngleFromEnglishDescription(phase: String?): Int? {
            return if (phase.isNullOrEmpty()) {
                null
            } else when (phase.lowercase(Locale.getDefault())) {
                "waxingcrescent", "waxing crescent" -> 45
                "first", "firstquarter", "first quarter" -> 90
                "waxinggibbous", "waxing gibbous" -> 135
                "full", "fullmoon", "full moon" -> 180
                "waninggibbous", "waning gibbous" -> 225
                "third", "thirdquarter", "third quarter", "last", "lastquarter", "last quarter" -> 270
                "waningcrescent", "waning crescent" -> 315
                else -> 360
            }
        }
    }
}
